### PR TITLE
feat!: migrate to fastify org

### PR DESCRIPTION
**Pull request by Eomm**
still need test to improve coverage

feel free to commit on this branch 👍🏽

### CODE REVIEW

**Comment by Uzlopak** [comment]
Can you add this package to fastify/plugins team? I have no write permissions

**Comment by Eomm** [comment]
fixed 👍🏽

**Comment by Uzlopak** [comment]
I thought about the plugin.

First of all currently it is called reply.eh for early hints. we should probably rename that to make the attribute self explaining. 

Second I think the implementation should be more configured through routeOptions instead of setting the entryHints in the handler. Maybe add an onRouteHandler which makes it possible to set earlyHints per Route.

This brings us to the third issue: formatEntry is currently called on every request. So if we pass a EarlyHintItem it gets serialized every time. So if we actually serialize only once, we can avoid alot of redundant computations.  So setting earlyHints as routeOption would make it possible to serialize the earlyHints once.

**Comment by climba03003** [comment]
1. We should update and change the implementation to support `headers`.
See https://github.com/nodejs/node/pull/44180#issuecomment-1219641389

2. Update `replyDecorator` and use `this` pointer to `reply`. There's no need of wrapper function and `onRequest` hook.

> Second I think the implementation should be more configured through routeOptions instead of setting the entryHints in the handler.

The current implementation is perfectly fine. `route options` means we lose flexibility to customize and calling `early hints` multiple times.

**Comment by Eomm** [comment]
I would track all the new features (such headers) to dedicated issues:

- support headers
- decorator name customization
- flow/performance optimization for `formatEntry`
- automatic metadata plugin (https://github.com/fastify/fastify/issues/2683#issuecomment-1219440127)

This PR accomplish its target and we can focus on which issues must be done before releasing this version.

**Comment by Uzlopak** [comment]
Well. I think I did my part then :). I trust my code changes. So maybe somebody else should review it?!

If we use the writeEarlyHints function we need to use combined test coverage of different test runs as we support node 14 but early hints will most likely only be backported to latest node 16.

The default workflow does not support combined test coverage.

File: index.js [hunk 1]
```diff
@@ -76,11 +29,7 @@ function fastifyEH (fastify, opts, next) {
       },
       add: function (content) {
         const p = new Promise((resolve) => {
-          if (reply.raw.socket) {
```
Comment (zekth) [inline index.js:79]: Help me remember, this is a Fastify 3 way right?

Comment (Uzlopak) [reply index.js:79]: I reverted this. Description below

**Comment by Uzlopak** [comment]
I reverted that part regarding reply.raw.socket.write. 

Unit tests are green if only use reply.raw.socket.write. But If I use the example (which I provided) and autocannon it, it gets internal server errors. So under pressure the sockets are sometimes null. 

So i reverted that change as it seems that it fixes that 500 errors.

But still it smells. 

```sh
uzlopak@Battlestation:~/workspace/fastify-early-hints$ autocannon http://localhost:3000 --renderStatusCodes
Running 10s test @ http://localhost:3000
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 0 ms  │ 0 ms │ 0.21 ms │ 2.98 ms │ 49 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬───────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%   │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼───────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 5243    │ 5243    │ 32559 │ 44063   │ 31407.4 │ 10341.5 │ 5242    │
├───────────┼─────────┼─────────┼───────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 2.04 MB │ 2.04 MB │ 13 MB │ 17.7 MB │ 12.6 MB │ 4.18 MB │ 2.04 MB │
└───────────┴─────────┴─────────┴───────┴─────────┴─────────┴─────────┴─────────┘
┌──────┬────────┐
│ Code │ Count  │
├──────┼────────┤
│ 103  │ 95958  │
├──────┼────────┤
│ 200  │ 218097 │
└──────┴────────┘

Req/Bytes counts sampled once per second.
# of samples: 10

218097 2xx responses, 95958 non 2xx responses
314k requests in 10.01s, 126 MB read
```

Imho: The amount of 103 codes should be the same as 200 codes. But they arent.

**Comment by zekth** [comment]
Are we sure autocannon if handling properly the 100-199 codes? Can you write a log line in a case of no possible to write to socket to ensure you get issues with the autocannon?

File: test/warn.test.js [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  fastify.get('/', (req, reply) => {
+    reply.eh.add([
+      { href: '/style.css', rel: 'preload', as: 'style', cors: 'invalid' },
+      { href: '/script.js', rel: 'preload', as: 'script' }
+    ])
+    return { hello: 'world' }
+  })
+  fastify.listen({ port: 3001 }, (err) => {
+    t.error(err)
+    exec('curl -D - http://localhost:3001').then(({ stdout, stderr }) => {
```
Comment (zekth) [inline test/warn.test.js:85]: We will be able to replace this test using undici eg: https://github.com/nodejs/undici/blob/48c03b62fa9c7e69444fba4a936a7a5e5c88e144/test/client-request.js#L476-L499

**Comment by Uzlopak** [comment]
I think we should change the logic to


```js
  const serialize = function (message) {
    return `HTTP/1.1 103 Early Hints${CRLF}${message}${CRLF}`
  }

  function earlyHints (reply) {
    let message = ''

    return {
      close: function (done) {
        if (message) {
          if (reply.raw.socket) {
            reply.raw.socket.write(serialize(message), 'utf-8', done)
          } else {
            reply.raw.write(serialize(message), 'utf-8', done)
          }
        } else {
          done()
        }
      },
      add: function (content) {
        for (let i = 0; i < content.length; ++i) {
          message += `${formatEntry(content[i], formatEntryOpts)}${CRLF}`
        }
      }
    }
  }
```

So we would send only once 103 and not every time we do a .add-call. 

See https://chromium.googlesource.com/chromium/src/+/master/docs/early-hints.md#what_s-not-supported

> Chrome ignores the second and following Early Hints responses. Chrome only handles the first Early Hints response so that Chrome doesn't apply inconsistent security policies (e.g. Content-Security-Policy).

File: test/warn.test.js (hunk 2 above)
Comment (zekth) [reply test/warn.test.js:85]: Look the test closely, it adds in an `info` array

**Comment by climba03003** [comment]
> we would send only once 103 and not every time we do a .add-call.

We should follow the specification instead of browser behavior.
It is up to the `client` to decide how to handle `early hints`, but as a `server` we should not limit how it is sent.

**Comment by Uzlopak** [comment]
@mcollina 
Do you have any idea what the issue could be?

**Comment by Uzlopak** [comment]
I added a stress test with autocanonnen. So now the unit tests are green, because we cover everything.

@mcollina 
It seems that autocannon handles the 103 statuscode for the calculation of total amount of requests.

So for 10000 requests with autocannon: 
```sh
┌──────┬───────┐
│ Code │ Count │
├──────┼───────┤
│ 103  │ 3051  │
├──────┼───────┤
│ 200  │ 6949  │
└──────┴───────┘
```

3051+5949 = 10000

**Comment by Uzlopak** [comment]
I personally would expect 5000 103 and 5000 200.

**Comment by Uzlopak** [comment]
So I would actually recommend to modify the actual functionality. 

So instead of having add and close, we should decoreReply with 'writeEarlyHints' 

```js
  fastify.decoreReply('writeEarlyHints', async function () {
    await new Promise(resolve => {
      if (reply.raw.socket) {
        reply.raw.socket.write(serialize(content), 'utf-8', resolve)
      } else {
        reply.raw.write(serialize(content), 'utf-8', resolve)
      }
    })
  })
```

So what happens then is that we have to await the writeEarlyHints in the handler, as the close function would not await to write the early hints. 

Also I think the node implementation differs from our implementation. We have cors, but node also has some other fields like title or anchor
https://github.com/wingleung/node/blob/c204024fcaa97ac74270f4270ce69c9dccd9a1dd/lib/internal/validators.js#L262

**Comment by Uzlopak** [comment]
@zekth 

nodejs is using ascii instead of utf-8 to write to socket. Should we also switch to ascii?

**Comment by zekth** [comment]
> nodejs is using ascii instead of utf-8 to write to socket. Should we also switch to ascii?

I guess so. Good catch.

**Comment by Eomm** [comment]
> 3051+5949 = 10000

The count shows 6949, 11k 😖

**Comment by Uzlopak** [comment]
@Eomm 
@climba03003 
@zekth 

What should we do now?

The codebase has 100% code coverage and is with no breaking change (except fastify 4.0 requirement).

Should we merge this and then have some more follow up PRs?  Or should we do everything in this PR?

I would prefer the first. We can merge this and then we can have a PR regarding
- [ ] use ascii instead of utf-8 and if we need some more checks
- [ ] use of native writeEarlyHints
- [ ] use reply decorator instead of two hooks

**Comment by Uzlopak** [comment]
I think autocannon does not work correctly as it mixes sometimes 103 with 200. If I put the example.js under stress test with autocannon -f and then run a curl request it is always correct order etc..

**Comment by Uzlopak** [comment]
creating follow up PRs

**Pull request merged**